### PR TITLE
TensorIterator to RNN/GRU/LSTM Sequence transformation

### PR DESCRIPTION
**Pull request by itikhono**
currently in disabled state

### CODE REVIEW

File: inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/convert_ti_to_sequences.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include <memory>
+
+#include <transformations_visibility.hpp>
+
+#include <ngraph/pass/graph_rewrite.hpp>
+
+namespace ngraph {
+namespace pass {
+
+    class TRANSFORMATIONS_API ConvertTensorIteratorToLSTMSequence;
```
Comment (ilyachur) [inline inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/convert_ti_to_sequences.h:20]: Looks like the code styli is not aligned with other transformations

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp [hunk 2]
```diff
@@ -23,14 +23,12 @@ ngraph::pass::ConvertLSTMSequenceMatcher::ConvertLSTMSequenceMatcher() {
             return false;
         }
 
-        const auto& W = std::dynamic_pointer_cast<ngraph::opset4::Constant>(
-                lstm_sequence->input_value(4).get_node_shared_ptr());
+        const auto& W = lstm_sequence->input_value(4).get_node_shared_ptr();
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp:26]: Why did you remove dynamic pointer cast?

File: inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/convert_ti_to_sequences.cpp [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+        auto input_B = std::make_shared<ngraph::opset4::Constant>(ngraph::element::f32, ngraph::Shape{4});
+
+        auto cell = std::make_shared<ngraph::opset4::LSTMCell>(input_data, input_H_state, input_C_state,
+                                                               input_W, input_R, input_B, 1);
+
+        auto axis_unsqueeze = std::make_shared<ngraph::opset4::Constant>(ngraph::element::i64, ngraph::Shape{1}, 1);
+        auto unsqueeze = std::make_shared<ngraph::opset4::Unsqueeze>(cell, axis_unsqueeze);
+        ngraph::pattern::Matcher matcher(unsqueeze);
+
+        bool flag = false;
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/convert_ti_to_sequences.cpp:45]: nit: can you change the name? the "flag" name doesn't help to understand the purpose of this flag.

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp (hunk 2 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp:26]: in TI to Sequence transformation, we insert an additional Unsqueeze layer, so we have here W->Unsqueeze, not Constant

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp [hunk 4]
```diff
@@ -23,14 +23,12 @@ ngraph::pass::ConvertLSTMSequenceMatcher::ConvertLSTMSequenceMatcher() {
...
         if (!W) {
             return false;
         }
 
-        const auto& R = std::dynamic_pointer_cast<ngraph::opset4::Constant>(
-                lstm_sequence->input_value(5).get_node_shared_ptr());
+        const auto& R = lstm_sequence->input_value(5).get_node_shared_ptr();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp:31]: If this nodes are used as inputs you need to avoid of get_node_shared_ptr call.

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp [hunk 5]
```diff
@@ -84,14 +82,12 @@ ngraph::pass::ConvertGRUSequenceMatcher::ConvertGRUSequenceMatcher() {
             return false;
         }
 
-        auto W = std::dynamic_pointer_cast<ngraph::opset4::Constant>(
-                gru_sequence->input_value(3).get_node_shared_ptr());
+        auto W = gru_sequence->input_value(3).get_node_shared_ptr();
         if (!W) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp:86]: Looks like this checks are no more needed.

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp (hunk 4 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp]: fixed

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp (hunk 5 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp]: removed

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp [hunk 6]
```diff
@@ -146,27 +134,23 @@ ngraph::pass::ConvertRNNSequenceMatcher::ConvertRNNSequenceMatcher() {
             return false;
         }
 
-        auto W = std::dynamic_pointer_cast<ngraph::opset4::Constant>(
-                rnn_sequence->input_value(3).get_node_shared_ptr());
-        if (!W) {
+        // Bidirectional cases are not supported
+        // todo: add exception
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp:138]: TODO? Why cannot you add exception in this PR?

**Review by ilyachur — APPROVED** [review]
In general LGTM

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp (hunk 6 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp:138]: I don't see any exceptions in any transformation. As I understand the logic of this: if the transformation was not successful (return false), then theoretically there is an opportunity to apply another transformation or other way to manage this situation, but exception broke this.

Comment (ilyachur) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp:138]: In this case please remove todo. From my point of view it is OK, that we return false in case of error.
We shouldn't throw exception in this case.

Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_sequences_to_sequences_ie.cpp]: fixed
